consultation_analyser/consultations/ml_pipeline.py

```diff
@@ -0,0 +1,104 @@
+from uuid import UUID
+from typing import List, NamedTuple, Dict, Union
+
+from umap.umap_ import UMAP
+from sentence_transformers import SentenceTransformer
+from hdbscan import HDBSCAN
+from bertopic import BERTopic
+from bertopic.vectorizers import ClassTfidfTransformer
+from sklearn.feature_extraction.text import CountVectorizer
+import numpy as np
+import pandas as pd
+
+from consultation_analyser.consultations import models
+
+
+RANDOM_STATE = 12  # For reproducibility
+
+
+def get_embeddings_for_question(
+    answers_list: List[Dict[str, Union[UUID, str]]], embedding_model_name: str = "thenlper/gte-small"
+) -> List[Dict[str, Union[UUID, str, np.ndarray]]]:
+    free_text_responses = [answer["free_text"] for answer in answers_list]
+    embedding_model = SentenceTransformer(embedding_model_name)
+    embeddings = embedding_model.encode(free_text_responses)
+    z = zip(answers_list, embeddings)
+    answers_list_with_embeddings = [dict(list(d.items()) + [("embedding", embedding)]) for d, embedding in z]
+    return answers_list_with_embeddings
+
+
+def get_topic_model(answers_list_with_embeddings: List[Dict[str, Union[UUID, str, np.ndarray]]]) -> BERTopic:
+    free_text_responses_list = [answer["free_text"] for answer in answers_list_with_embeddings]
+    embeddings_list = [answer["embedding"] for answer in answers_list_with_embeddings]
+    embeddings = np.array(embeddings_list)
+    # Set random_state so that we can reproduce the results
+    umap_model = UMAP(n_neighbors=15, n_components=5, min_dist=0.0, metric="cosine", random_state=RANDOM_STATE)
+    hdbscan_model = HDBSCAN(
+        min_cluster_size=3, metric="euclidean", cluster_selection_method="eom", prediction_data=True
+    )
+    vectorizer_model = CountVectorizer(stop_words="english")
+    ctfidf_model = ClassTfidfTransformer()
+    topic_model = BERTopic(
+        umap_model=umap_model, hdbscan_model=hdbscan_model, vectorizer_model=vectorizer_model, ctfidf_model=ctfidf_model
+    )
+    topic_model.fit_transform(free_text_responses_list, embeddings=embeddings)
+    return topic_model
+
+
+def get_answers_and_topics(topic_model: BERTopic, answers_list: List[Dict[str, Union[UUID, str]]]) -> pd.DataFrame:
+    # Answers free text/IDs need to be in the same order
+    free_text_responses = [answer["free_text"] for answer in answers_list]
+    answers_id_list = [answer["id"] for answer in answers_list]
+    # Assign topics to answers
+    answers_df = topic_model.get_document_info(free_text_responses)
+    answers_df["id"] = answers_id_list
+    answers_df = answers_df[["id", "Topic", "Name", "Representation"]]
+    return answers_df
+
+
+class AnswerRow(NamedTuple):
+    id: int
+    Name: str
+    Representation: List
+
+
+def save_theme_to_answer(question: models.Question, answer_row: AnswerRow) -> models.Answer:
```

Comment:
v minor point but this is a purely ORM based function, so does it belong in `models.py` , perhaps it could a method of `Answer`?

Reply:
I'm undecided - let's see how I feel in the morning.

If it was a method in answer, would probably be of the form:
```
def save_theme(self, question, answer_row):
    # rest of code similar
```
Can't decide if neater or not.

Reply:
Did change this - turned out a lot neater in the end (now we don't need to pass through question parameter).